### PR TITLE
Implement InstrumentationLibrary translation to jaeger

### PR DESCRIPTION
**Pull request by iNikem**
Implements [missing part of OTLP->Jaeger translation](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/sdk_exporters/jaeger.md#instrumentationlibrary)

### CODE REVIEW

**Comment by iNikem** [comment]
Would like to ask for help with failed "Loadtest" job. I am not familiar with it

File: translator/trace/jaeger/traces_to_jaegerproto.go
Comment (bogdandrutu) [inline translator/trace/jaeger/traces_to_jaegerproto.go:84]: maps are more inefficient than slices, consider to return a slice. or simply pass the InstrumentationLibrary to the spanToJaegerProto (will probably be more efficient than maps).

**Comment by iNikem** [comment]
> You need to add the conversion from Jaeger -> OTLP as well

Oh :( This means I have to take a slice of spans and group them by InstrumentationLibrary? Hmmm... let me see if my Go-fu is strong enough...

**Comment by iNikem** [comment]
> You need to add the conversion from Jaeger -> OTLP as well

Hopefully done

**Comment by iNikem** [comment]
@open-telemetry/collector-approvers Gently asking for some reviews

**Comment by iNikem** [comment]
Not sure how my changes could cause these build failures...

File: translator/trace/jaeger/jaegerproto_to_traces.go
Comment (dmitryax) [inline translator/trace/jaeger/jaegerproto_to_traces.go:144]: As @bogdandrutu mentioned, maps are pretty inefficient. This is a hot spot for conversion and most of the times we will allocate a map with just one key. So it would be great to avoid allocating a map if possible. I think it would be acceptable if we break down the spans like this:
```
{jSpansWithLibrary1..., jSpansWithLibrary2..., jSpansWithLibrary1...} -> {otlpLibrary1Spans, otlpLibrary2Spans, otlpLibrary1Spans}
```
It should simplify things and do not require a map.

Comment (dmitryax) [reply translator/trace/jaeger/jaegerproto_to_traces.go:144]: @bogdandrutu any thought on this ^ ?

Comment (iNikem) [reply translator/trace/jaeger/jaegerproto_to_traces.go:144]: That will duplicate library info on each span, increasing memory usage. Do we have performance tests or benchmarks for that?

Comment (dmitryax) [reply translator/trace/jaeger/jaegerproto_to_traces.go:144]: It may increase memory usage for the worst cases when we have many spans with different libraries jumping like spansLib1, spansLib2, spansLib1 etc. I don't know how realistic it that use case? I assume that in 99.9% cases it'll be just one library, right? Otherwise we might also consider sorting by library before, but not sure if we need to optimize for that.
I don't think we have existing performance tests for this, but it might worth adding a benchmark test to this particular function along with the change.

Comment (iNikem) [reply translator/trace/jaeger/jaegerproto_to_traces.go:144]: Mm, probably I misunderstood your proposal. If I remove the map, then I will need to convert every `model.Span` to a pair of `(instrumentationLibrary;pdata.Span)`, right? Which means having a separate copy of `instrumentationLibrary` for every span.

Comment (iNikem) [reply translator/trace/jaeger/jaegerproto_to_traces.go:144]: My other concern is the scope of this PR, which is again exploding. I just wanted to implement recent changes in OTLP->Jaeger translation regarding `InstrumentationLibrary`. It already grew into including Jaeger->OTLP translation which is much more complicated. If I have also to add a benchmark test for this, that will blow my time budget for this task too much.

Comment (dmitryax) [reply translator/trace/jaeger/jaegerproto_to_traces.go:144]: I understand, but and the same time we shouldn't be introducing performance issues with this change just to merge it sooner. You don't need to return pairs, probably nothing should be returned as it was before the change, you can just append to an `InstrumentationLibrary` slice provided to this function instead.

File: translator/trace/jaeger/traces_to_jaegerproto.go
Comment (dmitryax) [inline translator/trace/jaeger/traces_to_jaegerproto.go:251]: Does't seem like we need `libraryTagsFound`. the condition can be removed

File: translator/trace/jaeger/traces_to_jaegerproto.go
Comment (dmitryax) [inline translator/trace/jaeger/traces_to_jaegerproto.go:460]: This function can be simplified and just return `[]model.KeyValue`

File: translator/trace/jaeger/traces_to_jaegerproto.go
Comment (dmitryax) [inline translator/trace/jaeger/traces_to_jaegerproto.go:463]: `return nil`

File: translator/trace/jaeger/jaegerproto_to_traces.go
Comment (iNikem) [reply translator/trace/jaeger/jaegerproto_to_traces.go:144]: I afraid I don't understand you. This function takes an array of `model.Span`. Before, it converted them one by one and added them all into a slice of `pdata.Span`. Now we have to extract `pdata.InstrumentationLibrary` from every `model.Span` _in addition_ to `pdata.Span`. Then all those spans have to be grouped by `pdata.InstrumentationLibrary` into a collection of `pdata.InstrumentationLibrarySpans`. Each instance of `pdata.InstrumentationLibrarySpans` contains `pdata.InstrumentationLibrary` and a slice of `pdata.Span` created by that `pdata.InstrumentationLibrary`. Without the map of `pdata.InstrumentationLibrary -> collection of pdata.Span` I don't know how perform "group by" operation.

File: translator/trace/jaeger/traces_to_jaegerproto.go
Comment (dmitryax) [inline translator/trace/jaeger/traces_to_jaegerproto.go:461]: this should be moved below `if il.IsNil() {` check.

File: translator/trace/jaeger/jaegerproto_to_traces.go
Comment (dmitryax) [reply translator/trace/jaeger/jaegerproto_to_traces.go:144]: I suggest the following:
1. Sort jaeger spans slice by instrumentation library (optional);
2. Iterate over the jaeger spans slice and build `pdata.InstrumentationLibrarySpans` structure on the fly: 
      - if jaeger instrumentation library is the same as previous: add the span to the current `pdata.InstrumentationLibrarySpans->Spans` slice.
      - otherwise: add new item to `pdata.InstrumentationLibrarySpans` slice.

My initial point that I wouldn't care about duplicated instrumentation library data, that's why I think `1` is optional.

Does it make sense?

Comment (iNikem) [reply translator/trace/jaeger/jaegerproto_to_traces.go:144]: A, yes, it does :) If you think this extra sorting is faster overall than using a map, I will try that tomorrow.

Comment (dmitryax) [reply translator/trace/jaeger/jaegerproto_to_traces.go:144]: Yes, I think it must be faster, especially assuming that in vast majority cases it'll be just one library.

I'd say it's still an open question if we need the sorting at all. There are plans to remove this instrumentation library wrapping sometime in future. I'd like to hear what others think about it. cc @bogdandrutu

Comment (bogdandrutu) [reply translator/trace/jaeger/jaegerproto_to_traces.go:144]: Indeed, maps are expensive but if we do what you suggest @dmitryax we will allocate tons of data, so in this particular case I tend to agree with the map approach.

We will allocate a new "instrumentation library object", a new slice of spans, and a new "entry to the resource slice of instr lib". So 3 extra allocations which I think is more than the overhead of using maps.

Comment (dmitryax) [reply translator/trace/jaeger/jaegerproto_to_traces.go:144]: Yeah, that's the case with no sorting and different instrumentation libraries. I still think this is going to be unnecessary map allocation just for one key most of the times, but probably we should not predict it too much and assume the worth case. Happy to close the discussion if there are no other comments

File: translator/trace/jaeger/traces_to_jaegerproto.go
Comment (iNikem) [reply translator/trace/jaeger/traces_to_jaegerproto.go:463]: I followed the example of existing `getTagsFromTraceState` function. And in fact all other functions in this file return some object and a boolean. I can do as you say, no problem, but isn't consistency important? :)

Comment (dmitryax) [reply translator/trace/jaeger/traces_to_jaegerproto.go:463]: I did't look at other functions :) this one seems to be cleaner if simplified, but the consistency is a valid point. feel free to keep it as is

**Review by dmitryax — APPROVED** [review]
LGTM

**Comment by iNikem** [comment]
@dmitryax @open-telemetry/collector-maintainers Should I do anything else before this PR can be merged?

**Comment by dmitryax** [comment]
We need to ask maintainers to merge it. cc @bogdandrutu

**Comment by iNikem** [comment]
Is build failure related to my changes? Does not look so :(

**Comment by dmitryax** [comment]
@iNikem looks like `two-instrumentation-libraries` is randomly failing because the instrumentation libraries order cannot be predicted after grouping with the map. You'd need to change the test to make sure it doesn't prescribe the order

**Comment by iNikem** [comment]
@open-telemetry/collector-maintainers @open-telemetry/collector-approvers can you review, approve and merge? :)

**Review by dmitryax — APPROVED** [review]

**Comment by tigrannajaryan** [comment]
@bogdandrutu are you OK with the current approach with maps? OK to merge or you want to have another look?

**Review by bogdandrutu — APPROVED** [review]

**Pull request merged**
